cedar-lean/Cedar/Thm/Data/List.lean

```diff
@@ -770,31 +797,135 @@ theorem mapM_head_tail {α β γ} {f : α → Except β γ} {x : α} {xs : List
   cases h₂ : mapM' f xs <;>
   simp [h₂, pure, Except.pure]
 
+/-! ### Forall₂ -/
+
+theorem forall₂_implies_all_left {α β} {R : α → β → Prop} {xs : List α} {ys : List β} :
+  List.Forall₂ R xs ys →
+  ∀ x ∈ xs, ∃ y ∈ ys, R x y
+:= by
+  intro h
+  induction h
+  case nil =>
+    simp only [not_mem_nil, false_and, exists_false, imp_self, implies_true]
+  case cons xhd yhd xtl ytl hhd _ ih =>
+    intro x hx
+    simp only [mem_cons] at hx
+    rcases hx with hx | hx
+    case inl =>
+      subst hx
+      exists yhd
+      simp only [mem_cons, true_or, hhd, and_self]
+    case inr =>
+      have ⟨y, ih⟩ := ih x hx
+      exists y
+      simp only [mem_cons, ih, or_true, and_self]
+
+theorem forall₂_implies_all_right {α β} {R : α → β → Prop} {xs : List α} {ys : List β} :
+  List.Forall₂ R xs ys →
+  ∀ y ∈ ys, ∃ x ∈ xs, R x y
+:= by
+  intro h
+  induction h
+  case nil =>
+    simp only [not_mem_nil, false_and, exists_false, imp_self, implies_true]
+  case cons xhd yhd xtl ytl hhd _ ih =>
+    intro y hy
+    simp only [mem_cons] at hy
+    rcases hy with hy | hy
+    case inl =>
+      subst hy
+      exists xhd
+      simp only [mem_cons, true_or, hhd, and_self]
+    case inr =>
+      have ⟨x, ih⟩ := ih y hy
+      exists x
+      simp only [mem_cons, ih, or_true, and_self]
+
 /-! ### mapM' -/
 
-theorem mapM'_some_implies_all_some {α β} {f : α → Option β} {xs : List α} {ys : List β} :
+theorem mapM'_ok_implies_forall₂ {α β γ} {f : α → Except γ β} {xs : List α} {ys : List β} :
+  List.mapM' f xs = .ok ys →
+  List.Forall₂ (λ x y => f x = .ok y) xs ys
+:= by
+  intro h₁
+  induction xs generalizing ys
+  case nil =>
+    simp only [mapM'_nil, pure, Except.pure, Except.ok.injEq] at h₁
+    subst h₁
+    exact List.Forall₂.nil
+  case cons xhd xtl ih =>
+    simp only [mapM'_cons, pure, Except.pure] at h₁
+    cases h₂ : f xhd <;>
+    simp only [h₂, Except.bind_err, Except.bind_ok] at h₁
+    rename_i yhd
+    cases h₃ : mapM' f xtl <;>
+    simp only [h₃, Except.bind_err, Except.bind_ok] at h₁
+    rename_i ytl
+    simp only [Except.ok.injEq] at h₁
+    subst h₁
+    exact List.Forall₂.cons h₂ (ih h₃)
+
+theorem mapM'_ok_implies_all_ok {α β γ} {f : α → Except γ β} {xs : List α} {ys : List β} :
+  List.mapM' f xs = .ok ys →
+  ∀ x, x ∈ xs → ∃ y, y ∈ ys ∧ f x = .ok y
```

Comment:
Same suggestion from `forall₂_implies_all_*` above. (Assuming the syntax is ok -- I haven't compiled locally.)
```suggestion
  ∀ x ∈ xs, ∃ y ∈ ys, f x = .ok y
```